pkg/gatewayserver/redis/registry.go

```diff
@@ -57,29 +60,71 @@ func (r *GatewayConnectionStatsRegistry) Set(ctx context.Context, ids *ttnpb.Gat
 	defer trace.StartRegion(ctx, "set gateway connection stats").End()
 
 	uk := r.key(uid)
-	if stats == nil {
-		err = r.Redis.Del(ctx, uk).Err()
-	} else {
-		err = ttnredis.LockedWatch(ctx, r.Redis, uk, lockerID, r.LockTTL, func(tx *redis.Tx) error {
-			pb := &ttnpb.GatewayConnectionStats{}
-			if err := ttnredis.GetProto(ctx, tx, uk).ScanProto(pb); err != nil && !errors.IsNotFound(err) {
+	err = ttnredis.LockedWatch(ctx, r.Redis, uk, lockerID, r.LockTTL, func(tx *redis.Tx) error {
+		stored := &ttnpb.GatewayConnectionStats{}
+		cmd := ttnredis.GetProto(ctx, tx, uk)
+		if err := cmd.ScanProto(stored); errors.IsNotFound(err) {
+			stored = nil
+		} else if err != nil {
+			return err
+		}
+
+		var pb *ttnpb.GatewayConnectionStats
+		if stored != nil {
+			pb = &ttnpb.GatewayConnectionStats{}
+			if err := cmd.ScanProto(pb); err != nil {
 				return err
 			}
-			if err := pb.SetFields(stats, paths...); err != nil {
+			if pb, err = applyGatewayConnectionStatsFieldMask(nil, pb, gets...); err != nil {
 				return err
 			}
-			_, err := ttnredis.SetProto(ctx, tx, uk, pb, ttl)
+		}
+
+		var sets []string
+		pb, sets, err = f(pb)
+		if err != nil {
 			return err
-		})
-	}
+		}
+		if stored == nil && pb == nil {
+			return nil
+		}
+		var pipelined func(redis.Pipeliner) error
+		if pb == nil {
+			pipelined = func(p redis.Pipeliner) error {
+				p.Del(ctx, uk)
+				return nil
+			}
+		} else {
```

Comment:
This branch wouldn't be used since these we never really delete stats but good to keep it consistent.